File: csharp/MapleStory.OpenAPI/Src/TMS/Dto/Character/CharacterHyperStat.cs
using Newtonsoft.Json;
using Base = MapleStory.OpenAPI.Common.DTO;

namespace MapleStory.OpenAPI.TMS.DTO
{
    /// <summary>
    /// 角色極限屬性資訊
    /// </summary>
    public class CharacterHyperStatDTO : Base.CharacterHyperStatDTO<CharacterHyperStatPresetDTO>
    {
        /// <summary>
        /// 要搜尋的日期 (TST，每日資料中的小時與分鐘將顯示為 0)
        /// </summary>
        [JsonProperty("date")]
        public override DateTimeOffset? Date
        {
            get
            {
                return _date?.ToOffset(TimeSpan.FromHours(8));
            }
            set
            {
                _date = value;
            }
        }

        private DateTimeOffset? _date;

        /// <summary>
        /// 角色職業
        /// </summary>
        [JsonProperty("character_class")]
        public override string? CharacterClass { get; set; }

        /// <summary>
        /// 目前套用的預設編號
        /// </summary>
        [JsonProperty("use_preset_no")]
        public override string? UsePresetNo { get; set; }

        /// <summary>
        /// 可用的極限屬性點數上限
        /// </summary>
        [JsonProperty("use_available_hyper_stat")]
        public override int? UseAvailableHyperStat { get; set; }

        /// <summary>
        /// 預設 1 的極限屬性資訊
        /// </summary>
        [JsonProperty("hyper_stat_preset_1")]
        public override List<CharacterHyperStatPresetDTO> HyperStatPreset1 { get; set; }

        /// <summary>
        /// 預設 1 剩餘的極限屬性點數
        /// </summary>
        [JsonProperty("hyper_stat_preset_1_remain_point")]
        public override int? HyperStatPreset1RemainPoint { get; set; }

        /// <summary>
        /// 預設 2 的極限屬性資訊
        /// </summary>
        [JsonProperty("hyper_stat_preset_2")]
        public override List<CharacterHyperStatPresetDTO> HyperStatPreset2 { get; set; }

        /// <summary>
        /// 預設 2 剩餘的極限屬性點數
        /// </summary>
        [JsonProperty("hyper_stat_preset_2_remain_point")]
        public override int? HyperStatPreset2RemainPoint { get; set; }

        /// <summary>
        /// 預設 3 的極限屬性資訊
        /// </summary>
        [JsonProperty("hyper_stat_preset_3")]
        public override List<CharacterHyperStatPresetDTO> HyperStatPreset3 { get; set; }

        /// <summary>
        /// 預設 3 剩餘的極限屬性點數
        /// </summary>
        [JsonProperty("hyper_stat_preset_3_remain_point")]
        public override int? HyperStatPreset3RemainPoint { get; set; }
    }

    /// <summary>
    /// 預設的極限屬性資訊
    /// </summary>
    public class CharacterHyperStatPresetDTO : Base.CharacterHyperStatPresetDTO
    {
        /// <summary>
        /// 能力值類型
        /// </summary>
        [JsonProperty("stat_type")]
        public override string StatType { get; set; }

        /// <summary>
        /// 投入於能力值的點數
        /// </summary>
        [JsonProperty("stat_point")]
        public override long? StatPoint { get; set; }

        /// <summary>
        /// 能力值等級
        /// </summary>
        [JsonProperty("stat_level")]
        public override long StatLevel { get; set; }

        /// <summary>
        /// 能力值提升量
        /// </summary>
        [JsonProperty("stat_increase")]
        public override string? StatIncrease { get; set; }
    }
}

File: python/maplestory_openapi/api/tms/dto/character/character_hyper_stat.py
from datetime import datetime
from pydantic import BaseModel, field_validator

from maplestory_openapi.api.common.dto.character.character_hyper_stat import CharacterHyperStatPreset as BaseCharacterHyperStatPreset
from maplestory_openapi.api.common.dto.character.character_hyper_stat import CharacterHyperStat as BaseCharacterHyperStat


class CharacterHyperStatPreset(BaseModel, BaseCharacterHyperStatPreset):
    """
    預設的極限屬性資訊

    Attributes:
        stat_type (str): 能力值類型
        stat_point (int or None): 投入於能力值的點數
        stat_level (int): 能力值等級
        stat_increase (str or None): 能力值提升量
    """
    stat_type: str
    stat_point: int | None
    stat_level: int
    stat_increase: str | None


class CharacterHyperStat(BaseModel, BaseCharacterHyperStat):
    """
    角色極限屬性資訊

    Attributes:
        date (datetime or None): 要搜尋的日期 (TST，每日資料中的小時與分鐘將顯示為 0)
        character_class (str): 角色職業
        use_preset_no (str): 目前套用的預設編號
        use_available_hyper_stat (int): 可用的極限屬性點數上限
        hyper_stat_preset_1 (list[CharacterHyperStatPreset]): 預設 1 的極限屬性資訊
        hyper_stat_preset_1_remain_point (int): 預設 1 剩餘的極限屬性點數
        hyper_stat_preset_2 (list[CharacterHyperStatPreset]): 預設 2 的極限屬性資訊
        hyper_stat_preset_2_remain_point (int): 預設 2 剩餘的極限屬性點數
        hyper_stat_preset_3 (list[CharacterHyperStatPreset]): 預設 3 的極限屬性資訊
        hyper_stat_preset_3_remain_point (int): 預設 3 剩餘的極限屬性點數
    """
    date: datetime | None
    character_class: str
    use_preset_no: str
    use_available_hyper_stat: int
    hyper_stat_preset_1: list[CharacterHyperStatPreset]
    hyper_stat_preset_1_remain_point: int
    hyper_stat_preset_2: list[CharacterHyperStatPreset]
    hyper_stat_preset_2_remain_point: int
    hyper_stat_preset_3: list[CharacterHyperStatPreset]
    hyper_stat_preset_3_remain_point: int

    @field_validator(
        "hyper_stat_preset_1",
        "hyper_stat_preset_2",
        "hyper_stat_preset_3",
        mode="before"
    )
    @classmethod
    def null_as_empty(cls, v):
        if v is None:
            return []
        return v
